### PR TITLE
Feat/add net interface resource object

### LINKED ISSUE #168: PHP error Invalid argument supplied for foreach()
### Steps to reproduce
1. Click on system information

### Expected behaviour
That no error will get logged.

### Actual behaviour
On every call this error will be logged:
`"message":"Invalid argument supplied for foreach() at /var/www/html/owncloud/apps/serverinfo/lib/OperatingSystems/DefaultOs.php#120"`

### Server configuration

**Operating system:** CentOS 7.7

**Web server:** Apache

**Database:** MariaDB

**PHP version:** 7.2

**Nextcloud version:** 17.0.2

**Updated from an older Nextcloud/ownCloud or fresh install:**
Updated from older

**Where did you install Nextcloud from:**
From the nextcloud web site

**Signing status:**
<details>
<summary>Signing status</summary>
```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - activity: 2.10.1
  - bruteforcesettings: 1.4.0
  - calendar: 1.7.1
  - cloud_federation_api: 1.0.0
  - dav: 1.13.0
  - dicomviewer: 1.2.1
  - federatedfilesharing: 1.7.0
  - files: 1.12.0
  - files_pdfviewer: 1.6.0
  - files_rightclick: 0.15.1
  - files_sharing: 1.9.0
  - files_videoplayer: 1.6.0
  - gallery: 18.4.0
  - logreader: 2.2.0
  - lookup_server_connector: 1.5.0
  - nextcloud_announcements: 1.6.0
  - notifications: 2.5.0
  - oauth2: 1.5.0
  - password_policy: 1.7.0
  - piwik: 0.5.3
  - privacy: 1.1.0
  - provisioning_api: 1.7.0
  - recommendations: 0.5.0
  - serverinfo: 1.7.0
  - sharebymail: 1.7.0
  - socialsharing_email: 1.0.6
  - support: 1.0.1
  - text: 1.1.1
  - theming: 1.8.0
  - twofactor_backupcodes: 1.6.0
  - twofactor_u2f: 5.0.1
  - updatenotification: 1.7.0
  - viewer: 1.2.0
  - workflowengine: 1.7.0
Disabled:
  - accessibility
  - admin_audit
  - comments
  - encryption
  - federation
  - files_external
  - files_trashbin
  - files_versions
  - firstrunwizard
  - survey_client
  - systemtags
  - user_ldap

```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "speicher.terrortux.de"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "17.0.2.1",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "forcessl": true,
        "maxZipInputSize": 83886080,
        "allowZipDownload": true,
        "mail_smtpmode": "smtp",
        "mail_smtphost": "***REMOVED SENSITIVE VALUE***",
        "mail_smtpport": "25",
        "log_type": "file",
        "loglevel": "1",
        "logfile": "data\/nextcloud.log",
        "logtimezone": "Europe\/Berlin",
        "log_authfailip": true,
        "theme": "",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "mail_from_address": "***REMOVED SENSITIVE VALUE***",
        "forceSSLforSubdomains": true,
        "maintenance": false,
        "asset-pipeline.enabled": true,
        "updatechecker": false,
        "htaccess.RewriteBase": "\/",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "filelocking.enabled": true,
        "memcache.locking": "\\OC\\Memcache\\APCu",
        "mysql.utf8mb4": true,
        "overwrite.cli.url": "https:\/\/XXXXXXXXX",
        "updater.release.channel": "stable"
    }
}

```
</details>

**Are you using external storage, if yes which one:** local/smb/sftp/...
no

**Are you using encryption:** yes/no
no

**Are you using an external user-backend, if yes which one:** LDAP/ActiveDirectory/Webdav/...
no
```

### PR DESCRIPTION
**Pull request by kesselb**
Close #168

### CODE REVIEW

**Comment by szaimen** [comment]
Sorry, I don't feel able to review this :/

**Comment by kesselb** [comment]
> Sorry, I don't feel able to review this :/

No problem ;)

File: lib/OperatingSystems/DefaultOs.php [hunk 1]
```diff
@@ -141,50 +146,55 @@ public function getNetworkInfo(): array {
...
-						$iface['speed'] = $iface['speed'] . ' Mbps';
-					}
-				} else {
-					$iface['speed'] = 'unknown';
+		foreach ($this->getNetInterfaces() as $interfaceName => $interface) {
+			$netInterface = new NetInterface();
+			$data[] = $netInterface;
+
+			$netInterface->setName($interfaceName);
+			$netInterface->setUp($interface['up']);
```
Comment (come-nc) [inline lib/OperatingSystems/DefaultOs.php:156]: Add name and up as constructor parameters maybe?

File: lib/OperatingSystems/DefaultOs.php [hunk 2]
```diff
@@ -141,50 +146,55 @@ public function getNetworkInfo(): array {
...
+			if ($netInterface->isLoopback()) {
+				continue;
+			}
+
+			$interfacePath = '/sys/class/net/' . $interfaceName;
+
+			try {
+				$netInterface->setMAC($this->readContent($interfacePath . '/address'));
+
+				$speed = (int)$this->readContent($interfacePath . '/speed');
```
Comment (come-nc) [inline lib/OperatingSystems/DefaultOs.php:176]: Any chance this can be a float on 32bits?

File: lib/Resources/NetInterface.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+	/**
+	 * @return string[]
+	 */
+	public function getIPv6(): array {
+		return $this->ipv6;
+	}
+
+	public function addIPv6(string $ipv6): void {
+		$this->ipv6[] = $ipv6;
+		if ($ipv6 === '127.0.0.1') {
```
Comment (st3iny) [inline lib/Resources/NetInterface.php:77]:
```suggestion
		if ($ipv6 === '::1') {
```

File: tests/lib/DefaultOsTest.php [hunk 4]
```diff
@@ -207,4 +208,42 @@ public function testGetDiskInfoInvalidCommandOutput(): void {
 	public function testSupported(): void {
 		$this->assertTrue($this->os->supported());
 	}
+
+	public function testGetNetworkInterfaces(): void {
+		$this->os->method('getNetInterfaces')
+			->willReturn(include __DIR__ . '/../data/linux_net_get_interfaces.php');
```
Comment (come-nc) [inline tests/lib/DefaultOsTest.php:214]: This weird, there is no better way? Maybe json_decode a json file instead?

File: tests/lib/FreeBSDTest.php [hunk 5]
```diff
@@ -97,54 +98,46 @@ public function testGetMemoryNoData(): void {
...
-				}
 				if ($command === '/sbin/ifconfig pflog0') {
 					return file_get_contents(__DIR__ . '/../data/freebsd_interface_pflog0');
 				}
 				if ($command === '/sbin/ifconfig epair0b') {
 					return file_get_contents(__DIR__ . '/../data/freebsd_interface_epair0b');
 				}
-
-				// Regex matches way more than the interface names, so if it doesn't match any of the defined ones, throw.
-				throw new \RuntimeException();
```
Comment (come-nc) [inline tests/lib/FreeBSDTest.php:116]: Removing this means other commands are executed? Which ones? I’d feel better if the test lists them and throw for the rest.

**Review by st3iny — APPROVED** [review]
Looks good so far and works. Did you test in on FreeBSD?

File: lib/Resources/NetInterface.php (hunk 3 above)
Comment (come-nc) [inline lib/Resources/NetInterface.php:77]: `::1`, no?
Also same question as above, are the IP normalized in any way?

File: lib/OperatingSystems/DefaultOs.php (hunk 1 above)
Comment (kesselb) [reply lib/OperatingSystems/DefaultOs.php:156]: Like the idea, thanks!

File: lib/OperatingSystems/DefaultOs.php (hunk 2 above)
Comment (kesselb) [reply lib/OperatingSystems/DefaultOs.php:176]: Not within the next 10 years.

File: lib/Resources/NetInterface.php [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+	/**
+	 * @return string[]
+	 */
+	public function getIPv4(): array {
+		return $this->ipv4;
+	}
+
+	public function addIPv4(string $ipv4): void {
+		$this->ipv4[] = $ipv4;
+		if ($ipv4 === '127.0.0.1') {
```
[the remark this answers is not in the document]
Comment (kesselb) [reply lib/Resources/NetInterface.php:63]: I hope ;) 
https://www.php.net/manual/en/function.inet-ntop.php is used internally to convert the ip address.

File: lib/Resources/NetInterface.php (hunk 3 above)
Comment (kesselb) [reply lib/Resources/NetInterface.php:77]: :see_no_evil:

Comment (kesselb) [reply lib/Resources/NetInterface.php:77]: :see_no_evil:

File: templates/settings-admin.php [hunk 7]
```diff
@@ -192,25 +199,25 @@ function FormatMegabytes(int $byte): string {
...
-									<?php if (!empty($interface['mac'])): ?>
+									<span class="info"><?php p($interface->getSpeed()) ?> (<?php p($l->t('Duplex:') . ' ' . $interface->getDuplex()) ?>)</span><br>
+									<?php if (!empty($interface->getMAC())): ?>
 										<?php p($l->t('MAC:')); ?>
-										<span class="info"><?php p($interface['mac']) ?></span><br>
+										<span class="info"><?php p($interface->getMAC()) ?></span><br>
 									<?php endif; ?>
 									<?php p($l->t('IPv4:')); ?>
-									<span class="info"><?php p($interface['ipv4']) ?></span><br>
+									<span class="info"><?= implode(', ', $interface->getIPv4()); ?>
```
[the remark this answers is not in the document]
Comment (kesselb) [reply templates/settings-admin.php:217]: Thanks :+1:

File: tests/lib/FreeBSDTest.php (hunk 5 above)
Comment (kesselb) [reply tests/lib/FreeBSDTest.php:116]: Added the RuntimeException back.

File: tests/lib/DefaultOsTest.php (hunk 4 above)
Comment (kesselb) [reply tests/lib/DefaultOsTest.php:214]: Good question.

`getNetInterfaces` is a wrapper for `net_get_interfaces`  to mock the network interfaces for testing. linux_net_get_interfaces.php is the result of var_export(net_get_interfaces());

Right now, it's as close as it gets to the original return. 
It should work to export as json and decode for the test again.  

What do you prefer?

**Comment by kesselb** [comment]
> Did you test in on FreeBSD?

Not really. 
I run `php -r "var_export(net_get_interfaces());"` to check if `net_get_interfaces` is available and capture the data for the tests on a virtual FreeBSD machine.

File: tests/lib/DefaultOsTest.php (hunk 4 above)
Comment (come-nc) [reply tests/lib/DefaultOsTest.php:215]: I would feel better with json, yes. This requires to go read the PHP file to check what it does. A json_decode call will be clearer.

File: lib/OperatingSystems/DefaultOs.php [hunk 8]
```diff
@@ -141,50 +146,52 @@ public function getNetworkInfo(): array {
...
+				if ($speed >= 1000) {
+					$netInterface->setSpeed($speed / 1000 . ' Gbps');
```
Comment (come-nc) [inline lib/OperatingSystems/DefaultOs.php:175]: How is translation handled for these units? Is it?

Comment (kesselb) [reply lib/OperatingSystems/DefaultOs.php:175]: Gbps and Mpbs are not translated.
I guess that's okay for most people as noone complained afaik ;)

File: tests/lib/DefaultOsTest.php (hunk 4 above)
Comment (kesselb) [reply tests/lib/DefaultOsTest.php:215]: Okay.

File: lib/OperatingSystems/DefaultOs.php (hunk 8 above)
Comment (come-nc) [reply lib/OperatingSystems/DefaultOs.php:175]: Not ideal, but that was already like that before the PR?

File: tests/lib/DefaultOsTest.php (hunk 4 above)
Comment (kesselb) [reply tests/lib/DefaultOsTest.php]: Done.

**Review by come-nc — APPROVED** [review]

**Pull request merged**